bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/ServiceAccounts/ServiceAccountAuthorizationHandler.cs

```diff
@@ -0,0 +1,106 @@
+
+namespace Bit.Commercial.Core.SecretsManager.AuthorizationHandlers.ServiceAccounts;
+
+public class
+    ServiceAccountAuthorizationHandler : AuthorizationHandler<ServiceAccountOperationRequirement, ServiceAccount>
+{
+    private readonly ICurrentContext _currentContext;
+    private readonly IServiceAccountRepository _serviceAccountRepository;
+    private readonly IUserService _userService;
+
+    public ServiceAccountAuthorizationHandler(ICurrentContext currentContext, IUserService userService,
+        IServiceAccountRepository serviceAccountRepository)
+    {
+        _currentContext = currentContext;
+        _userService = userService;
+        _serviceAccountRepository = serviceAccountRepository;
+    }
+
+    protected override async Task HandleRequirementAsync(AuthorizationHandlerContext context,
+        ServiceAccountOperationRequirement requirement,
+        ServiceAccount resource)
+    {
+        if (!_currentContext.AccessSecretsManager(resource.OrganizationId))
+        {
+            return;
+        }
+
+        switch (requirement)
+        {
+            case not null when requirement == ServiceAccountOperations.Create:
+                await CanCreateServiceAccountAsync(context, requirement, resource);
+                break;
+            case not null when requirement == ServiceAccountOperations.Read:
+                await CanReadServiceAccountAsync(context, requirement, resource);
+                break;
+            case not null when requirement == ServiceAccountOperations.Update:
+                await CanUpdateServiceAccountAsync(context, requirement, resource);
+                break;
+            default:
+                throw new ArgumentException("Unsupported project operation requirement type provided.",
+                    nameof(requirement));
+        }
+    }
+
+    private async Task CanCreateServiceAccountAsync(AuthorizationHandlerContext context,
+        ServiceAccountOperationRequirement requirement, ServiceAccount resource)
+    {
+        var (accessClient, _) = await GetAccessClientAsync(context, resource.OrganizationId);
+        var hasAccess = accessClient switch
+        {
+            AccessClientType.NoAccessCheck => true,
+            AccessClientType.User => true,
+            AccessClientType.ServiceAccount => false,
+            _ => false,
+        };
+
+        if (hasAccess)
+        {
+            context.Succeed(requirement);
+        }
+    }
+
+    private async Task CanReadServiceAccountAsync(AuthorizationHandlerContext context,
+        ServiceAccountOperationRequirement requirement, ServiceAccount resource)
+    {
+        var (accessClient, userId) = await GetAccessClientAsync(context, resource.OrganizationId);
+        var access =
+            await _serviceAccountRepository.AccessToServiceAccountAsync(resource.Id, userId,
+                accessClient);
+
+        if (access.Write)
+        {
+            context.Succeed(requirement);
+        }
+    }
```

Comment:
Note this read checks write access, which is counterintuitive.

Since this is just a refactor, I opted to keep the original functionality, which is checking for write access.

```
     if (serviceAccount == null)
        {
            throw new NotFoundException();
        }

        if (!_currentContext.AccessSecretsManager(serviceAccount.OrganizationId))
        {
            throw new NotFoundException();
        }

        var orgAdmin = await _currentContext.OrganizationAdmin(serviceAccount.OrganizationId);
        var accessClient = AccessClientHelper.ToAccessClient(_currentContext.ClientType, orgAdmin);

        var hasAccess = accessClient switch
        {
            AccessClientType.NoAccessCheck => true,
            AccessClientType.User => await _serviceAccountRepository.UserHasWriteAccessToServiceAccount(id, userId),
            _ => false,
        };

        if (!hasAccess)
        {
            throw new NotFoundException();
        }
```

If we would like to swap this to read in this PR, I don't have any objections.

Reply:
I ended up changing to correctly checking read access in this PR https://github.com/bitwarden/server/pull/2869/commits/3f412d8f183b5479f8a26f7890057aaaf5911063

Doesn't seem like much of a scope increase for this PR.